include/envoy/http/async_client.h

```diff
@@ -38,7 +38,40 @@ class AsyncClient {
   };
 
   /**
-   * An in-flight HTTP request
+   * Notifies caller of async HTTP stream status.
+   */
+  class StreamCallbacks {
+  public:
+    virtual ~StreamCallbacks() {}
+
+    /**
+     * Called when the async HTTP stream receives headers.
+     * @param headers the headers received
+     * @param end_stream whether the response is header only
+     */
+    virtual void onHeaders(HeaderMapPtr&& headers, bool end_stream) PURE;
+
+    /**
+     * Called when the async HTTP stream receives data.
+     * @param data the data received
+     * @param end_stream whether the data is the last data frame
+     */
+    virtual void onData(Buffer::Instance& data, bool end_stream) PURE;
```

Comment:
How is stream closure signalled even if no data was delivered? Seems cleaner to separate into a separate callback otherwise closure handling will end up being overloaded into each handle impl